src/Analyzers/CSharp/CodeFixes/UseUTF8StringLiteral/UseUTF8StringLiteralCodeFixProvider.cs

```diff
@@ -0,0 +1,191 @@
+
+        public override Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            RegisterCodeFix(context, CSharpAnalyzersResources.Use_UTF8_string_literal, nameof(CSharpAnalyzersResources.Use_UTF8_string_literal));
+            return Task.CompletedTask;
+        }
+
+        protected override async Task FixAllAsync(
+            Document document, ImmutableArray<Diagnostic> diagnostics,
+            SyntaxEditor editor, CodeActionOptionsProvider options, CancellationToken cancellationToken)
+        {
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            foreach (var diagnostic in diagnostics)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                var node = diagnostic.Location.FindNode(getInnermostNodeForTie: true, cancellationToken);
+                var stringValue = GetUTF8StringValueForDiagnostic(semanticModel, diagnostic, cancellationToken);
+
+                // If we're replacing a byte array that is passed to a parameter array, not and an explicit array creation
+                // then node will be the ArgumentListSyntax that the implicit array creation is just a part of, so we have
+                // to handle that separately, as we can't just replace node with a string literal
+                //
+                // eg given a method:
+                //     M(string x, params byte[] b)
+                // our diagnostic would be reported on:
+                //     M("hi", [|1, 2, 3, 4|]);
+                // but node will point to:
+                //     M([|"hi", 1, 2, 3, 4|]);
+
+                if (node is BaseArgumentListSyntax argumentList)
+                {
+                    editor.ReplaceNode(node, CreateArgumentListWithUTF8String(argumentList, diagnostic.Location, stringValue));
+                }
+                else
+                {
+                    editor.ReplaceNode(node, CreateUTF8String(node, stringValue));
+                }
+            }
+        }
+
+        private static string GetUTF8StringValueForDiagnostic(SemanticModel semanticModel, Diagnostic diagnostic, CancellationToken cancellationToken)
+        {
+            // For computing the UTF8 string we need the original location of the array creation
+            // operation, which is stored in additional locations.
+            var location = diagnostic.AdditionalLocations[0];
+            var node = location.FindNode(getInnermostNodeForTie: true, cancellationToken);
+
+            var operation = semanticModel.GetRequiredOperation(node, cancellationToken);
+
+            // Because we get the location from an IOperation.Syntax, sometimes we have to look a
+            // little harder to get back from syntax to the operation that triggered the diagnostic
+            if (operation is not IArrayCreationOperation arrayOp)
```

Comment:
instead of doing these sorts of 'if' checks on this side, it can be nice for the analyzer to check, and then set *properties* i in the diagnostic so that the fixer needs to do no work but switch off exactly which case the analyzer says it is in.

Reply:
I don't think there are properties that can help here, as the code fix needs to find the right IOperation. Will try to find some time to talk offline about it, I might be missing something

Reply:
Okay, I think maybe I know what you mean. lmk.